crates/sui-core/src/authority/backpressure.rs

```diff
@@ -0,0 +1,314 @@
+    // checkpoint. Otherwise, backpressure might prevent construction of the next checkpoint.
+    fn is_backpressure_suppressed(&self) -> bool {
+        self.certified <= self.executed
+    }
+}
+
+pub struct BackpressureManager {
+    // holds the executed and certified checkpoint watermarks.
+    // builder can run ahead of executor, or vice versa, so these watermarks
+    // do not have a strict ordering.
+    //
+    // We want to engage backpressure whenever builder is running ahead of the executor.
+    watermarks_sender: watch::Sender<Watermarks>,
+
+    // used by the WritebackCache to notify us when it has too many pending transactions in memory.
+    backpressure_sender: watch::Sender<bool>,
+}
+
+pub struct BackpressureSubscriber {
+    mgr: Arc<BackpressureManager>,
+}
+
+impl BackpressureManager {
+    pub fn new_for_tests() -> Arc<Self> {
+        Self::new_from_watermarks(Default::default())
+    }
+
+    fn new_from_watermarks(watermarks: Watermarks) -> Arc<Self> {
+        let (watermarks_sender, _) = watch::channel(watermarks);
+        let (backpressure_sender, _) = watch::channel(false);
+        Arc::new(Self {
+            watermarks_sender,
+            backpressure_sender,
+        })
+    }
+
+    pub fn new_from_checkpoint_store(store: &CheckpointStore) -> Arc<Self> {
+        let executed = store
+            .get_highest_executed_checkpoint_seq_number()
+            .expect("read cannot fail")
+            .unwrap_or_default();
+        let certified = store
+            .get_highest_synced_checkpoint_seq_number()
+            .expect("read cannot fail")
+            .unwrap_or_default();
+        info!(
+            ?executed,
+            ?certified,
+            "initializing backpressure manager from checkpoint store"
+        );
+        Self::new_from_watermarks(Watermarks {
+            executed,
+            certified,
+        })
+    }
+
+    pub fn update_highest_certified_checkpoint(&self, seq: CheckpointSequenceNumber) {
+        self.watermarks_sender.send_if_modified(|watermarks| {
+            if seq > watermarks.certified {
+                debug_assert_eq!(seq, watermarks.certified + 1);
+                watermarks.certified = seq;
+                debug!(?watermarks, "updating highest certified checkpoint");
+                true
+            } else {
+                false
+            }
+        });
+    }
+
+    pub fn update_highest_executed_checkpoint(&self, seq: CheckpointSequenceNumber) {
+        self.watermarks_sender.send_if_modified(|watermarks| {
+            if seq > watermarks.executed {
+                debug_assert_eq!(seq, watermarks.executed + 1);
+                watermarks.executed = seq;
+                debug!(?watermarks, "updating highest executed checkpoint");
+                true
+            } else {
+                false
+            }
+        });
+    }
+
+    pub fn set_backpressure(&self, backpressure: bool) {
```

Comment:
Maybe have a metric on last backpressure value, and / or counter on the number of times backpressure is set to false / true.